src/log.h

```diff
@@ -185,35 +188,155 @@ class LogOutputBuffer : public ICombinedLogOutput {
 	Logger &m_logger;
 };
 
+class LogTarget {
+public:
+	// Must be thread-safe. These can be called from any thread.
+	virtual bool hasOutput() = 0;
+	virtual void log(const std::string &buf) = 0;
+};
 
-extern StreamLogOutput stdout_output;
-extern StreamLogOutput stderr_output;
-extern std::ostream null_stream;
 
-extern std::ostream *dout_con_ptr;
-extern std::ostream *derr_con_ptr;
-extern std::ostream *derr_server_ptr;
+class StreamProxy {
+public:
+	StreamProxy(std::ostream *os) : m_os(os) { }
+
+	template<typename T>
+	StreamProxy& operator<<(T&& arg) {
```

Comment:
what about `const T&` types?

Reply:
This is perfect forwarding, it's a special C++ category. It's not the same thing as rvalue reference.
As the name suggests, it will forward the type of T exactly, including the lvalue/rvalue category.